---
title: Java 异常
tags:
 - Java SE
categories:
 - Java 基础
---


### 异常类

- Error：系统错误，程序无法处理。
- Exception：程序运行时出现的错误，程序可以处理。

Error 和 Exception 都是 Throwable 的子类，Throwable、Error、Exception 都是存放在 java.lang 包中。

Error 常见的子类：VirtualMachineError、AWTError、IOError。

VirtualMachineError 的子类：StackOverflowError、OutOfMemoryError，用来描述内存溢出等系统问题。

Exception 常见的子类：IOException 和 RuntimeException

IOException 存放在 java.io 包中，RuntimeException 存放在 java.lang 包中。

IOException 常见的子类：FileLockInterruptionException、FileNotFoundException、FilerException，这些异常通常都是处理通过 IO 流进行文件传输的时候发生的错误。

RuntimeException 常见的子类：

- ArithmeticException：表示数学运算异常。
- ClassNotFoundException：表类未定义异常。
- IllelArgumentException：表示参数格式错误。
- ArrayIndexOutOfBounds：表示数组下标越界。
- NullPointException：表示空指针异常。
- NoSuchMethodException：表示方法未定义异常。
- NumberFormatException：表示将其他数据类型转为数值类型发生的类型不匹配异常。



### throw 和 throws

throw 和 throws 是 Java 在处理异常时使用的两个关键字，使用的方式和表示的含义完全不同。

Java 中抛出异常有 3 种方式：

- try-catch
  
- 使用 throw 是开发者主动抛出异常，即读到 throw 代码就一定抛出异常，基本语法：throw new Exception()，是一种基于代码的逻辑而主动抛出异常的方式。

- try-catch 和 throw 都是作用于具体的逻辑代码，throws 是作用于方法的，用来描述方法可能会抛出的异常。

如果方法 throws 的是 RuntimeException 异常或者其子类，外部调用时可以不处理，JVM 会处理。

如果方法 throws 的是 Exception 异常或者其子类，外部调用时必须处理，否则报错。

```java
public class Test {
	public static void main(String[] args) throws Exception { // 第二步：必须加throws Exception或者对test方法进行try-catch(因为下面这个方法有throws Exception(如果下面的方法自身try-catch处理则不用)，上层必须对它进行处理) 如果test方法throws RuntimeException则可以不加
		test("123");
	}
	
	public static void test(String str) throws Exception {
		throw new Exception(); // 第一步：主动抛出一个异常之后(相当于int i = 1 / 0) 必须在方面上面throws Exception或者对自己本身进行try-catch
	}
	
}
```

方法定义时的异常如果直接继承自 Exception，实际调用的时候需要手动处理（捕获异常/丢给虚拟机去处理）

方法定义时的异常如果继承自 RuntimeException，调用的时候不需要处理。



### 自定义异常

除了使用 Java 提供的异常外，也可以根据需求来自定义异常。

```java
public class MyNumberException extends RuntimeException {
	public MyNumberException(String error) {
		super(error);
	}
}
```

```java
public class Test {
	public static void main(String[] args){
		Test test = new Test();
		System.out.println(test.add("a"));
	}
	
	public int add(Object object){
		if(object instanceof Integer) {
			int num = (int)object;
			return ++num;
		}else {
			String error = "传入的参数不是整数类型";
			MyNumberException myNumberException = new MyNumberException(error);
			throw myNumberException;
		}
	}
}
```
